src/fldmod_by_const.jl

```diff
@@ -0,0 +1,147 @@
+"""
+    calculate_inv_coeff(::Type{T}, f)
+
+Returns the magic numbers needed for `fldmod_by_const(x, Val(C))`: `2^N / C`, or the
+`inv_coeff`. That `inv_coeff` has all leading zeros shifted away to maximize precision, so
+this also returns how many bits need to be shifted back to undo that optimization, making
+the return value: `((2^N / C) << toshift, toshift)`.
+"""
+# This function is marked `@pure` because it has no side-effects and depends on no global
+# state. The value will never change for a given (type,precision) pair.
+# This allows its result to be const-folded away when called with Const values.
+Base.@pure function calculate_inv_coeff(::Type{T}, f) where {T}
+    # First, calculate 2^nbits(T)/f
+    # We calculate it at double-precision and shift away leading zeros to preserve the most
+    # precision. Later, we will shift the final answer back to undo this operation.
+    UT = unsigned(T)
+    invcoeff = twoToTheSizeOf(UT)÷f
+    toshift = _leading_zeros(invcoeff)
+    invcoeff = invcoeff << toshift
+    # Now, truncate the result to sizeof(T), rounding to maintain precision.
+    invcoeff = _round_to_even(fldmod(invcoeff, typemax(UT))..., typemax(UT)) % T
+    return invcoeff, toshift
+end
+# These are needed to handle Int128, which widens to BigInt
+twoToTheSizeOf(::Type{T}) where {T} = typemax(widen(T))
+twoToTheSizeOf(::Union{Type{Int128}, Type{UInt128}}) = BigInt(2)^256
+twoToTheSizeOf(::Type{BigInt}) = BigInt(2)^256
+
+_leading_zeros(x) = leading_zeros(x)
+_leading_zeros(x::BigInt) = leading_zeros((x >> 128) % UInt128)
```

Comment:
Maybe a comment about how this is a special purpose leading-zeros for BigInt representing a 256-bit integer, and you know the leading 1 will be in the top 128 bits?

Reply:
Thanks for pointing this out. I meant to come back to this and try to rename it something more meaningful than "_leading_zeros". Do you have any suggestions?

I've added a comment as well, thanks.